policy: Generate L3-only filter also for rules with requirements.

jrajahalme opened the pull request:
FromRequires can be placed into the same rule as FromEndpoints, and
an L3-only filter needs to be generated also in that case.

Basically, as [From|To]Requires affects all rules selected for a given
identity, the processing of individual rules should not depend on the
(non)precense of [From|To]Requires in the same rule.

Fixes: 7321663ff85 ("policy: store L3-only policy in L4Filter")
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7894)
<!-- Reviewable:end -->

joestringer commented on pkg/policy/repository_test.go:
```diff
@@ -1902,8 +1923,7 @@ Resolving ingress policy for [any:bar]
 * Rule {"matchLabels":{"any:bar":""}}: selected
-    Enforcing requirements [{Key:any.baz Operator:In Values:[]}]
-      No label match for [any:foo]
+      Found all required labels
```
This was the part I was trying to avoid ;)

What does this mean?

(I mean, when you read the output, how do you interpret that this was actually a rule with only requirements and no allows? What does "Found all required labels" mean if the labels don't actually match the requirements?)

jrajahalme replied:
The problem here is that a rule with just `FromRequires` will never allow any traffic by itself. If we add the requirements (collected from this same rule) to this rule as MatchExpressions, then this rule will start allowing traffic from endpoints matching the requirement. Fix might be as simple as logging "Rule does not allow any traffic" when `fromEndpoints` is empty, instead of the "Found all required labels"?

jrajahalme replied:
Or just print out the "Found all required labels" when there are fromEndpoints to match?
That way this will be printed only if `fromEndpoints.Matches()` succeed`, which looks like the logic at other places where this is printed out?

jrajahalme replied:
I separated this fix to a new (first) commit, please take a look. Now a rule that does not allow anything will not print any trace output. Maybe we should print out a note that the "Rule does not allow any traffic" etc.?

jrajahalme committed:
policy: Report 'found all labels' only when 'Matches()' succeeds.

Remove adding requirements to rules without remote endpoint selectors
when tracing, as this would allow traffic based on the requirement
alone, which is against the policy language specification.

Move the trace print-out of "Found all required labels" to take place
only if an endpoint selector matches. This was already the case in
most places.

Fixes: b4f41cb ("policy: Rework ingress policy trace to L4PolicyMap")
Fixes: a41d8a0 ("policy: Rework egress policy trace to L4PolicyMap")
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Generate L3-only filter also for rules with requirements.

FromRequires can be placed into the same rule as FromEndpoints, and
an L3-only filter needs to be generated also in that case.

Basically, as [From|To]Requires affects all rules selected for a given
identity, the processing of individual rules should not depend on the
(non)precense of [From|To]Requires in the same rule.

Fixes: 7321663ff85 ("policy: store L3-only policy in L4Filter")
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer reviewed (approved):
I guess it's more clear by stating less. It's still not obvious when looking at requirements-only rules what exactly the policy trace is saying about it, but at least there isn't a misleading "found all labels" per the previous version. At some point we probably need to properly revisit the trace output for clarity.
